### PR TITLE
Added documentation for statistics service.

### PR DESCRIPTION
**Pull request by prasoon2211**
This PR addresses the fourth todo item (to fix docs and add new docs) in the PR #1725 

@Stibbons @sa2ajj @benallard Please have a look.

### CODE REVIEW

File: master/docs/developer/stats-service.rst [hunk 1]
```diff
@@ -15,14 +15,15 @@ Stats Service
 
 .. py:class:: StatsService
 
-   An instance of this class functions as a :class:`BuildbotService`. The instance of the running
-   service is initialized in the master configuration file (see :bb:cfg:`stats-service` for more).
+   An instance of this class functions as a :class:`BuildbotService`.
+   The instance of the running service is initialized in the master configuration file (see :bb:cfg:`stats-service` for more).
```
Comment (sa2ajj) [inline master/docs/developer/stats-service.rst]: `... for more information).`

**Comment by Stibbons** [comment]
+1

**Comment by sa2ajj** [comment]
We do not seem to have any guidelines written regarding documentation formatting, so I'll put a few notes here and then try to put it somewhere "central":

* when documenting a class, we usually put `.. py:class:: full.import.name.to.Class`.
  This might lead to clashes when use in different places, so the secondary locations, should use `:noindex:` option
* while documenting parameters the best result is produced when they are documented as
```
``parameter``
    description
``another``
    and a descirption for this one
```

I am not sure whether `**The default callback**` really deserves that kind of formatting.  I'll think about during the day and return back in the evening...

**Comment by prasoon2211** [comment]
@sa2ajj I've made changes according to your comment.

File: master/docs/developer/stats-service.rst [hunk 2]
```diff
@@ -13,16 +13,19 @@ Here is a diagram demonstrating the working of the stats service:
...
+   The instance of the running service is initialized in the master configuration file (see :bb:cfg:`stats-service` for more information).
    The running service is accessible everywhere in Buildbot via the :class:`BuildMaster`.
    The service is available at ``self.master.namedServices['<service-name>']``.
    It takes the following intialization arguments:
 
-   * ``storage_backends``: A list of storage backends. These are instance of subclasses of :class:`StatsStorageBase`.
-   * ``name``: The name of this service. This name can be used to access the running instance of this service using ``self.master.namedServices[name]``.
+   ``storage_backends``
+     A list of storage backends. These are instance of subclasses of :class:`StatsStorageBase`.
+   ``name``
```
Comment (sa2ajj) [inline master/docs/developer/stats-service.rst]: do you think you could update the previously provided docs for consistency?

I just looked at the rendered docs and it does not look very nice.

File: master/docs/developer/stats-service.rst [hunk 3]
```diff
@@ -68,44 +70,68 @@ Stats Service
...
-to aid in posting data by :class:`StatsService`
+Each storage backend has a Python client defined as part of :mod:`buildbot.statistics.storage_backends` to aid in posting data by :class:`StatsService`
 
 Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage backend.
 
-.. py:class:: StatsStorageBase
+.. py:class:: buildbot.statistis.storage_backends.base.StatsStorageBase
 
-   A base class for all storage services
+   A abstract base class for all storage services. It cannot be directly initialized - it would raise a ``TypeError`` otherwise.
```
Comment (sa2ajj) [inline master/docs/developer/stats-service.rst]: `An abstract class`

Do not forget one-line-per-sentence rule.

File: master/docs/developer/stats-service.rst [hunk 4]
```diff
@@ -68,44 +70,68 @@ Stats Service
...
 
-.. py:class:: InfluxStorageService
+   .. py:method:: thd_postStatsValue(self, post_data, series_name, context={})
+
+      :param post_data: A dict of key-value pairs that is sent for storage.
```
Comment (sa2ajj) [inline master/docs/developer/stats-service.rst]: Let's be consistent about parameter documentation.

Though Sphinx offers this way to document parameters, it takes way too much space.

File: master/docs/developer/stats-service.rst [hunk 5]
```diff
@@ -114,142 +140,416 @@ Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage back
...
-   * ``builder_name``: The name of builder whose times are to be recorded.
-   * ``callback=None``: The callback function for this class. See :class:`CaptureBuildTimes` for more.
+      This is an abstract method and needs to be implemented by all subclasses of this class.
+      This is a helper method to the ``consume`` method metioned above.
+      It checks whether a builder is allowed to send build times to the storage backend according to the configuration file.
+      It takes one argument:
+
+      ``builder_info``
+        (dict) The dictionary returned by the data API containing the builder information.
+
```
Comment (sa2ajj) [inline master/docs/developer/stats-service.rst]: there's no need for extra line here.

File: master/docs/developer/stats-service.rst [hunk 6]
```diff
@@ -13,16 +13,19 @@ Here is a diagram demonstrating the working of the stats service:
...
+   An instance of this class functions as a :class:`BuildbotService`.
+   The instance of the running service is initialized in the master configuration file (see :bb:cfg:`stats-service` for more information).
    The running service is accessible everywhere in Buildbot via the :class:`BuildMaster`.
    The service is available at ``self.master.namedServices['<service-name>']``.
    It takes the following intialization arguments:
 
-   * ``storage_backends``: A list of storage backends. These are instance of subclasses of :class:`StatsStorageBase`.
-   * ``name``: The name of this service. This name can be used to access the running instance of this service using ``self.master.namedServices[name]``.
+   ``storage_backends``
+     A list of storage backends. These are instance of subclasses of :class:`StatsStorageBase`.
```
Comment (sa2ajj) [inline master/docs/developer/stats-service.rst]: One sentence per line

File: master/docs/developer/stats-service.rst [hunk 7]
```diff
@@ -68,44 +70,68 @@ Stats Service
...
 Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage backend.
 
-.. py:class:: StatsStorageBase
+.. py:class:: buildbot.statistis.storage_backends.base.StatsStorageBase
 
-   A base class for all storage services
+   A abstract base class for all storage services. It cannot be directly initialized - it would raise a ``TypeError`` otherwise.
 
-.. py:class:: InfluxStorageService
+   .. py:method:: thd_postStatsValue(self, post_data, series_name, context={})
```
Comment (sa2ajj) [inline master/docs/developer/stats-service.rst]: actually, `{}` is dangerous as a default value.

File: master/docs/developer/stats-service.rst [hunk 8]
```diff
@@ -114,142 +140,416 @@ Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage back
...
+   .. py:method:: _builder_name_matches(self, builder_info):
+
+      This is an abstract method and needs to be implemented by all subclasses of this class.
+      This is a helper method to the ``consume`` method metioned above.
+      It checks whether a builder is allowed to send properties to the storage backend according to the configuration file.
+      It takes one argument:
+
+      ``builder_info``
+        (dict) The dictionary returned by the data API containing the builder information.
+
```
Comment (sa2ajj) [inline master/docs/developer/stats-service.rst]: No need for this extra empty line.

File: master/docs/developer/stats-service.rst [hunk 9]
```diff
@@ -68,44 +70,68 @@ Stats Service
 Storage backends
```
Comment (sa2ajj) [inline master/docs/developer/stats-service.rst]: No need for an extra empty line before this one.

File: master/docs/developer/stats-service.rst [hunk 10]
```diff
@@ -114,142 +140,416 @@ Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage back
...
+     .. py:function:: default_callback(props, property_name)
 
      It returns property value for ``property_name``.
+     It receives the following arguments:
+
+     ``props``
+       A dictionary of all build properties.
+     ``property_name``
+       Name of the build property to return.
+
```
Comment (sa2ajj) [inline master/docs/developer/stats-service.rst]: No need for this extra line.

File: master/docs/developer/stats-service.rst [hunk 11]
```diff
@@ -114,142 +140,416 @@ Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage back
...
+     A default callback is provided for this (see :class:`CapturePropertyBase` for more).
+   ``regex=False``
+     If this is set to ``True``, then the property name can be a regular expression.
+     All properties matching this regular expression will be sent for storage.
+
+   .. py:method:: _builder_name_matches(self, builder_info)
+
+      This method matches whether the ``builder_name`` used to initialize this class instance is the same as the builder for the curent build.
+      See :class:`CapturePropertyBase` for more information on this method.
+
```
Comment (sa2ajj) [inline master/docs/developer/stats-service.rst]: No need for this extra line.

File: master/docs/developer/stats-service.rst [hunk 12]
```diff
@@ -114,142 +140,416 @@ Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage back
...
+     The callback function that is used by ``CaptureProperty.consumer`` to post-process data before formatting it and sending it to the appropriate storage backends.
+     A default callback is provided for this (see :class:`CapturePropertyBase` for more).
+   ``regex=False``
+     If this is set to ``True``, then the property name can be a regular expression.
+     All properties matching this regular expression will be sent for storage.
+
+   .. py:method:: _builder_name_matches(self, builder_info)
+
+      Since this class is used to capture a property over all builders, it simply returns ``True`` in all cases which means that all builders are allowed to post the property set
+      in ``property_name`` to storage backends.
```
Comment (sa2ajj) [inline master/docs/developer/stats-service.rst]: One sentence per line (even if it's a long line).

File: master/docs/developer/stats-service.rst [hunk 13]
```diff
@@ -114,142 +140,416 @@ Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage back
...
 
    **The default callback:**
 
-      .. py:function:: default_callback
+      See ``CaptureBuildEndTime.__init__`` for the definition.
+
+   .. py:method:: _builder_name_matches(self, builder_info)
+
+      Returns ``True`` for all builders.
+
```
Comment (sa2ajj) [inline master/docs/developer/stats-service.rst]: No need for this extra line.

File: master/docs/developer/stats-service.rst [hunk 14]
```diff
@@ -114,142 +140,416 @@ Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage back
...
+   .. py:method:: _retValParams(self, msg)
 
    Returns a list containing one Python datetime object (end time) from ``msg`` dictionary.
 
+   .. py:method:: _builder_name_matches(self, builder_info)
+
+      This method matches whether the ``builder_name`` used to initialize this class instance is the same as the builder for the curent build.
+      See :class:`CaptureBuildTimes` for more information on this method.
+
```
Comment (sa2ajj) [inline master/docs/developer/stats-service.rst]: No need for this extra line.

File: master/docs/developer/stats-service.rst [hunk 15]
```diff
@@ -114,142 +140,416 @@ Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage back
...
+   .. py:method:: _builder_name_matches(self, builder_info):
+
+      This is an abstract method and needs to be implemented by all subclasses of this class.
+      This is a helper method to the ``consume`` method metioned above.
+      It checks whether a builder is allowed to send properties to the storage backend according to the configuration file.
+      It takes one argument:
+
+      ``builder_info``
+        (dict) The dictionary returned by the data API containing the builder information.
+
```
Comment (sa2ajj) [inline master/docs/developer/stats-service.rst]: No need for this extra line.

File: master/docs/developer/stats-service.rst [hunk 16]
```diff
@@ -114,142 +140,416 @@ Currently, only `InfluxDB <http://influxdb.com>`_ is supported as a storage back
...
+   **The default callback:**
+
+     See :class:`CaptureDataBase` of definition.
+
+   .. py:method:: _builder_name_matches(self, builder_info)
+
+      This method matches whether the ``builder_name`` used to initialize this class instance is the same as the builder for the curent build.
+      See :class:`CaptureBuildTimes` for more information on this method.
+
```
Comment (sa2ajj) [inline master/docs/developer/stats-service.rst]: No need for this extra line.

File: master/docs/manual/cfg-global.rst [hunk 17]
```diff
@@ -597,56 +603,177 @@ This is the main class for statistics service. It is initialized in the master c
...
+
+   ``builder_name``
+     The name of builder whose times are to be recorded.
+   ``report_in='seconds'``
+     Can be one of three: ``'seconds'``, ``'minutes'``, or ``'hours'``.
+     This is the units in which the build time will be reported.
+   ``callback=None``
+     (Optional) A custom callback function for this class.
+     This callback function should take in two Python datetime objects - a ``start_time`` and an ``end_time`` and return a string that will be sent for storage in the storage backends.
+
```
Comment (sa2ajj) [inline master/docs/manual/cfg-global.rst]: No need for these two extra lines.

File: master/docs/manual/cfg-global.rst [hunk 18]
```diff
@@ -597,56 +603,177 @@ This is the main class for statistics service. It is initialized in the master c
...
+   ``callback=None``
+     The callback function for this class.
+     This callback receives the data sent to  ``yieldMetricsValue`` as ``post_data`` (See :ref:`yieldMetricsValue`).
+     It must return a string that is to be sent to the storage backends for storage.
+
+
+.. py:class:: buildbot.statistics.capture.CaptureDataAllBuilders
+   :noindex:
+
+   Instance of this capture class for capturing arbitrary data that is not stored as build-data on all builders
```
Comment (sa2ajj) [inline master/docs/manual/cfg-global.rst]: Missing `.` at the end of the sentence.

File: master/docs/manual/cfg-global.rst [hunk 19]
```diff
@@ -597,56 +603,177 @@ This is the main class for statistics service. It is initialized in the master c
...
+     A dictionary of key value pair that is sent for storage.
+     The keys will act as columns in a database and the value is stored under that column.
+   ``buildid``
+     The integer build id of the current build.
+     Obtainable in all ``BuildSteps``.
 
 Along with using ``yieldMetricsValue``, the user will also need to use the ``CaptureData`` capture class.
+As an example, we can add the following to a build step:
+
+::
```
Comment (sa2ajj) [inline master/docs/manual/cfg-global.rst]: you might want to change it to

```
.. code-block:: python
```

we had quite a few issues when Sphinx/pygments was upgraded.

**Comment by sa2ajj** [comment]
Looks good, though:

* we need consistent way of documenting parameters (at least, within the same class/file)
* there's no need for two or three (or more!) empty lines between "parts"

File: master/docs/developer/stats-service.rst (hunk 2 above)
Comment (prasoon2211) [inline master/docs/developer/stats-service.rst]: I did change the whole thing though (including the previously uploaded docs). I will see if anything remains that I haven't corrected.

File: master/docs/developer/stats-service.rst (hunk 3 above)
Comment (prasoon2211) [inline master/docs/developer/stats-service.rst]: Missed this one. Will correct.

File: master/docs/developer/stats-service.rst (hunk 4 above)
Comment (prasoon2211) [inline master/docs/developer/stats-service.rst]: Okay.

File: master/docs/developer/stats-service.rst (hunk 2 above)
Comment (prasoon2211) [inline master/docs/developer/stats-service.rst]: @sa2ajj Are you asking me to change :class:``StatsStorageBase`` to :class:``buildbot.path.to.StatsStorageBase``?

Comment (sa2ajj) [inline master/docs/developer/stats-service.rst]: It's about `:param ...`, sorry for not being specific.

Comment (prasoon2211) [inline master/docs/developer/stats-service.rst]: Ah, okay.

**Comment by prasoon2211** [comment]
@sa2ajj I've addressed the issues that were present in the last commit. Please review.

**Comment by sa2ajj** [comment]
I built the docs locally and went through the rendered output.

Those "default callback" sections really confuse if they are not indented.

**Comment by sa2ajj** [comment]
@prasoon2211 any updates?

**Comment by prasoon2211** [comment]
I've been a bit busy with school work. I'll work on your suggestions this evening.

**Comment by prasoon2211** [comment]
@sa2ajj So, I've fixed the problems and rebased on top of master. Also, I checked the rendered docs in browser and everything looked good to me. Please review.

**Comment by sa2ajj** [comment]
Excellent.  Thank you

**Pull request merged**